### PR TITLE
Create an R interpreter for vulcan

### PR DESCRIPTION
**Pull request by dtm2451**
Following the path of how @coleshaw created an interpreter for `node`, this PR attempts to do the same for R.

Uses [jetpack](https://github.com/ankane/jetpack) for package management, which utilizes three files for configuration:
- `.Rprofile` - code that gets run on session start.  We can make use of that if needed.
- `renv.lock` - like a package.lock, holds info of installed package versions. (`renv` is an upstream package management tool that jetpack uses)
- `DESCRIPTION` - File used by the R package building system which renv/jetpack piggyback.  Can use this directly to add new deps by putting them into the `Imports:` field, and if not a CRAN package, the `Remotes:` field too.  OR to have jetpack take care of it, can also `make -C archimedes-r bash` `R` `jetpack::add("<package>")`.  See [jetpack docs](https://github.com/ankane/jetpack) for details.

Those three files are in the top-level of the `archimedes-r` folder.

### CODE REVIEW

File: archimedes-r/Makefile [hunk 1]
```diff
@@ -0,0 +1,9 @@
+include ../make-base/stubs.mk
+include ../make-base/docker-compose.mk
+
+app_name_capitalized:=$(shell echo ${app_name} | tr [a-z] [A-Z])
+app_service_name:=${app_name}_app
+fullTag:=$(IMAGES_PREFIX)$(app_name)$(IMAGES_POSTFIX)
+
+run-image-test:: compose-ready
+	docker run --rm -e "HOST_DIR=$${PWD}" -v "$${PWD}/tests/outputs:/app/tests/outputs" -v "archimedes-exec:/archimedes-exec" --mount 'type=bind,source=/var/run/docker.sock,target=/var/run/docker.sock,readonly' -e RELEASE_TEST=1 -e CI_SECRET=$${CI_SECRET} -e IS_CI=$${IS_CI} ${fullTag} /entrypoints/development.sh npm test
```
[the remark this answers is not in the document]
Comment (dtm2451) [reply archimedes-r/Makefile:9]: yes!  Good catch.  I didn't update this line yet compared to `npm`. Still need to figure out the testing side before finalizing.

**Comment by coleshaw** [comment]
This is awesome stuff! :100:

File: vulcan/config.yml.template [hunk 2]
```diff
@@ -29,8 +29,9 @@ default: &default
   :log_level: info
   :archimedes_run_image: etnaagent/archimedes
   :archimedes_interpreters:
-    :node: etnaagent/archimedes-node:production
-    :python: etnaagent/archimedes:production
+    :r: etnaagent/archimedes-r:latest
```
[the remark this answers is not in the document]
Comment (dtm2451) [reply vulcan/config.yml.template:32]: I'll test and try to figure out, but I've def noticed `:production` in errors I've gotten in dev when trying to `Update dependencies`.

Another open Q to come back to.

File: archimedes-r/docker-compose.yml [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+services:
+  archimedes-r_app:
+    image: archimedes-r
+    build:
+      context: ./
+      dockerfile: Dockerfile
+    volumes:
+      - .:/app:delegated
+      - archimedes-global-node-modules:/usr/local/lib/node_modules
```
[the remark this answers is not in the document]
Comment (dtm2451) [reply archimedes-r/docker-compose.yml:25]: Can do.  What's the purpose of having that cached?

[the remark this answers is not in the document]
Comment (coleshaw) [reply archimedes-r/docker-compose.yml:25]: Similar to how the other app images keep the dependencies in a Docker Volume (hence why you have to run `make update`), I think the volume keeps things a bit cleaner, and you won't see the installed libraries cluttering up your local drive. Also can then just mount it to the container without having to copy all the files **into** the container (so for things like npm, faster).

[the remark this answers is not in the document]
Comment (corps) [reply archimedes-r/docker-compose.yml:25]: Correct.  Although most of the performance gain comes from the fact that for node, we have several node_modules split between every project, and thus building into individual images would be fairly expensive.

For a single R image using a volume vs not is minor difference I suspect.

**Comment by dtm2451** [comment]
@coleshaw @corps Is there somewhere that we've already been documenting notes on how to add functions or packages to archimedes interpreters?  There are unique paths for each one.

File: vulcan/config.yml.template (hunk 2 above)
[the remark this answers is not in the document]
Comment (corps) [reply vulcan/config.yml.template:32]: The cnofig.yml.template only effects development.  Production configs are managed in the swarm and unaffected by this code.

**Comment by corps** [comment]
> @coleshaw @corps Is there somewhere that we've already been documenting notes on how to add functions or packages to archimedes interpreters? There are unique paths for each one.

Sorry, I actually don't understand the question.  Functions or packages to archimedes interpreters?  Unique paths?

**Comment by corps** [comment]
> I think this will also strongly relate to the general Vulcan workflow editing documentation and process ... which is unclear to me, right now. Perhaps a README.md in the root of each `archimedes-X` directory would work for now, since they are essentially standalone images?

Question:
1.  Are we building a standards platform --where we, the DL team, coordinate the correct implementations of standard methodologies, that we provide back to our DSs?--.  This is how our interpreters currently work -- they package the preset standard implementations and means of solving problems that we control and solve for them.  In return, in this design, it should not be assumed that 'editors' (we don't know who they are, but I assume they are someone not us) should be adding packages or altering images in significant ways without going through us.

2.  Are we building a generic tool that allows individual DS people to use tools they feel comfortable with to freely implement methodologies and publish them with frontends for easier tweaking and visualization?  In this world, we may publish some standard base images, but perhaps here Vulcan generally can accept --any docker or singularity image-- that follows the vulcan execution protocol (where to find inputs / outputs / scripts) a user wants to provide, and put the responsibility on them to design images with packages they wish to use already.

I think most vulcan design questions are currently unanswerable because we can't say "who" editors are or what use cases we're solving for that isn't just us doing things mostly.  I also worry we want a tool that is simultaneously flexible enough to fit any power user's workflow, as well as simple enough to fit our actual DS workflows, and be coordinated to contain standards tooling, while also open enough to contain every tooling someone might want.  (this, known as the kitchen sink, tends to fail on all accounts)

**Comment by dtm2451** [comment]
> > I think this will also strongly relate to the general Vulcan workflow editing documentation and process ... which is unclear to me, right now. Perhaps a README.md in the root of each `archimedes-X` directory would work for now, since they are essentially standalone images?
> 
> Question:
> 
> 1. Are we building a standards platform --where we, the DL team, coordinate the correct implementations of standard methodologies, that we provide back to our DSs?--.  This is how our interpreters currently work -- they package the preset standard implementations and means of solving problems that we control and solve for them.  In return, in this design, it should not be assumed that 'editors' (we don't know who they are, but I assume they are someone not us) should be adding packages or altering images in significant ways without going through us.
> 2. Are we building a generic tool that allows individual DS people to use tools they feel comfortable with to freely implement methodologies and publish them with frontends for easier tweaking and visualization?  In this world, we may publish some standard base images, but perhaps here Vulcan generally can accept --any docker or singularity image-- that follows the vulcan execution protocol (where to find inputs / outputs / scripts) a user wants to provide, and put the responsibility on them to design images with packages they wish to use already.
> 
> I think most vulcan design questions are currently unanswerable because we can't say "who" editors are or what use cases we're solving for that isn't just us doing things mostly. I also worry we want a tool that is simultaneously flexible enough to fit any power user's workflow, as well as simple enough to fit our actual DS workflows, and be coordinated to contain standards tooling, while also open enough to contain every tooling someone might want. (this, known as the kitchen sink, tends to fail on all accounts)

Agree with 1 as that's the primary direction that I think we're going in.  But I also expect that you're right that some requests towards the kitchen sink approach might come through.

I was thinking of documentation for us... I know how to make functions available through the python interpreter, and also now the R one as I've designed it.  But if one of you or Saurabh wanted to make code using a new R library, you'd need to ask me how to add it.

**Comment by corps** [comment]
> > > I think this will also strongly relate to the general Vulcan workflow editing documentation and process ... which is unclear to me, right now. Perhaps a README.md in the root of each `archimedes-X` directory would work for now, since they are essentially standalone images?
> > 
> > 
> > Question:
> > 
> > 1. Are we building a standards platform --where we, the DL team, coordinate the correct implementations of standard methodologies, that we provide back to our DSs?--.  This is how our interpreters currently work -- they package the preset standard implementations and means of solving problems that we control and solve for them.  In return, in this design, it should not be assumed that 'editors' (we don't know who they are, but I assume they are someone not us) should be adding packages or altering images in significant ways without going through us.
> > 2. Are we building a generic tool that allows individual DS people to use tools they feel comfortable with to freely implement methodologies and publish them with frontends for easier tweaking and visualization?  In this world, we may publish some standard base images, but perhaps here Vulcan generally can accept --any docker or singularity image-- that follows the vulcan execution protocol (where to find inputs / outputs / scripts) a user wants to provide, and put the responsibility on them to design images with packages they wish to use already.
> > 
> > I think most vulcan design questions are currently unanswerable because we can't say "who" editors are or what use cases we're solving for that isn't just us doing things mostly. I also worry we want a tool that is simultaneously flexible enough to fit any power user's workflow, as well as simple enough to fit our actual DS workflows, and be coordinated to contain standards tooling, while also open enough to contain every tooling someone might want. (this, known as the kitchen sink, tends to fail on all accounts)
> 
> Agree with 1 as that's the primary direction that I think we're going in. But I also expect that you're right that some requests towards the kitchen sink approach might come through.
> 
> I was thinking of documentation for us... I know how to make functions available through the python interpreter, and also now the R one as I've designed it. But if one of you or Saurabh wanted to make code using a new R library, you'd need to ask me how to add it.

Nah, I can figure it out, I promise.

**Comment by dtm2451** [comment]
To Do:
- [ ] testing via CI
- [ ] config.yml.template finalization
- [ ] Rlibs containerization
- [ ] investigate if can break a chunk of the R package installs, say Seurat & is 100+ deps, into a separate container which might not update as often, so that we can speed up `archimedes-r` rebuilds by starting with a `FROM <that-docker-image>`

**Comment by coleshaw** [comment]
> investigate if can break a chunk of the R package installs, say Seurat & is 100+ deps, into a separate container which might not update as often, so that we can speed up archimedes-r rebuilds by starting with a FROM <that-docker-image>

Curious -- is there an existing Seurat docker container, in Docker Hub? Or do the DS's have something you use on C4 (albeit perhaps a Singularity container)?

**Comment by dtm2451** [comment]
> > investigate if can break a chunk of the R package installs, say Seurat & is 100+ deps, into a separate container which might not update as often, so that we can speed up archimedes-r rebuilds by starting with a FROM
> 
> Curious -- is there an existing Seurat docker container, in Docker Hub? Or do the DS's have something you use on C4 (albeit perhaps a Singularity container)?

:+1:

There is a singularity container c4, and it benefits from Arjun aiming to have all our commonly used libraries built in.  Some extra bit and probably wonk that let's Rstudio-server work 1) with user-libs in each person's $HOME directory 2) over ssh. 

There also are Docker containers on DockerHub that have Seurat built in!  I'll test starting with one of those.

**Comment by graft** [comment]
> Agree with 1 as that's the primary direction that I think we're going in.

I'm not 100% sure about this. I often feel like the way archimedes is written now is too constrained, i.e., we are in tight control of what can be done, even as far as the python stdlib goes, and are stingy with what gets let through, can define our own arbitrary interfaces on top of other code (e.g., writing functions wrapping `magby` or `monoetna`), etc. This is all a somewhat stifling environment to imagine doing "editing" in, where the user will at least want to (a) be able to write their own shared libraries to reuse across workflows and (b) add packages they are interested in without much trouble.

The dichotomy between (1) and (2) is basically about what we think "without much trouble" means - how much and what is tolerable, who does it cause pain to, etc. Somehow we have to get it right, but a priori it is hard to know which approach of these two (or something else) will actually be less trouble.

**Comment by dtm2451** [comment]
With Cole's help, I moved the archimedes-r-base image creation to a separate repo on Thursday last week.  New steps for finalizing this path:
- [ ] Should the archimedes-r-base repo remain separate?
- [ ] path to copy the DESCRIPTION, renv.lock, and .Rprofile files out of the base image to check them in.
- [ ] reconfigure base image build to focus on those files and build off of the renv.lock
- [ ] path for :production vs :main/:development archimedes-r-base
- [ ] A last step of the PR: add additional packages

**Comment by dtm2451** [comment]
This needs to get pushed through!  A task for next week...
Tested today, and it still works!  Just need to double-check build system effects before merge:
- [ ] Ensure `jetpack::install()` step won't need to be run in `make up` and `make update` calls.  This step takes 20min+ so needs to be successfully passed over after cache pulls from etnaagent!

**Pull request merged**